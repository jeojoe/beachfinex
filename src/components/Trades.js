import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Map } from 'immutable';

import { actionCreators } from '../reducers/trades';

class Trades extends React.Component {
  componentDidMount() {
    const { initTrades, updateTrades } = this.props;
    const ws = new WebSocket('wss://api.bitfinex.com/ws/2');
    ws.onopen = () => {
      const openMsg = JSON.stringify({
        event: 'subscribe',
        channel: 'trades',
        symbol: 'tBTCUSD',
      });
      ws.send(openMsg);
    };
    ws.onmessage = (msg) => {
      // listen to 'te' for speed
      // http://blog.bitfinex.com/api/websocket-api-update/
      const data = JSON.parse(msg.data);
      if (data[1] && data[1] !== 'hb' && data[1] !== 'tu') {
        if (data[1] !== 'te') {
          initTrades(data[1]);
        } else {
          updateTrades(data[2]);
        }
      }
    };
    ws.onerror = err => console.log(err);
  }

  renderRow() {
    const { trades } = this.props;

    if (!trades) return 'Fetching..';

    return trades.map((order) => {
      const [id, timestamp, amount, price] = order;
      return (
        <div key={id}>
          <p>
            {timestamp}
            -
            {amount}
            -
            {price}
          </p>
        </div>
      );
    });
  }

  render() {
    return (
      <div>
        <div>{this.renderRow()}</div>
      </div>
    );
  }
}

Trades.propTypes = {
  trades: PropTypes.instanceOf(Map), // eslint-disable-line
  initTrades: PropTypes.func.isRequired,
  updateTrades: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    trades: state.trades,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    initTrades: orders => dispatch(actionCreators.initTrades(orders)),
    updateTrades: order => dispatch(actionCreators.updateTrades(order)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Trades);
