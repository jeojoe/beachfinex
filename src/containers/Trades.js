import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { List } from 'immutable';
import styled from 'styled-components';
import dayjs from 'dayjs';

import { actionCreators } from '../reducers/trades';
import HeaderRow from '../components/HeaderRow';

const ActionRow = styled.div`
  padding: 10px 0 0;
`;

const TradeRow = styled(HeaderRow)`
  & p {
    color: ${props => (props.amount >= 0 ? 'lime' : 'red')};
    font-size: 14px;
  }
`;

export class Trades extends React.Component {
  state = {
    subscribed: false,
    subscribing: true,
  };

  componentDidMount() {
    this.subscribe();
  }

  subscribe = () => {
    this.ws = new WebSocket('wss://api.bitfinex.com/ws/2');
    this.ws.onopen = this.onOpen;
    this.ws.onmessage = this.onMessage;
    this.ws.onerror = this.onError;
    this.setState({ subscribing: true });
  }

  unsubscribe = (err) => {
    this.ws.close();
    this.ws = null;
    this.setState({ subscribed: false });
    if (err) {
      // Has error, not user's action => subscribe again
      console.error(err);
      this.subscribe();
    }
  }

  toggleWebSocket = () => {
    if (this.ws) {
      this.unsubscribe();
    } else {
      this.subscribe();
    }
  }

  onOpen = () => {
    const openMsg = JSON.stringify({
      event: 'subscribe',
      channel: 'trades',
      symbol: 'tBTCUSD',
    });
    this.ws.send(openMsg);
  }

  onMessage = (msg) => {
    const { initTrades, updateTrades } = this.props;
    // listen to 'te' for speed
    // http://blog.bitfinex.com/api/websocket-api-update/
    const parsed = JSON.parse(msg.data);
    if (parsed.event === 'subscribed') {
      this.setState({ subscribed: true, subscribing: false });
      return;
    }
    const data = parsed;
    if (data[1] && data[1] !== 'hb' && data[1] !== 'tu') {
      if (data[1] !== 'te') {
        initTrades(data[1]);
      } else {
        updateTrades(data[2]);
      }
    }
  }

  onError = err => this.unsubscribe(err);

  renderRow() {
    const { trades } = this.props;

    if (!trades) return 'Fetching..';

    return trades.map(([id, timestamp, amount, price]) => (
      <TradeRow key={id} amount={amount}>
        <p>{dayjs(timestamp).format('HH:mm:ss')}</p>
        <p>{price}</p>
        <p>{Math.abs(amount)}</p>
      </TradeRow>
    ));
  }

  renderActions() {
    const { subscribed, subscribing } = this.state;
    return (
      <ActionRow>
        <b>{subscribed ? 'REAL-TIME' : 'OFFLINE' }</b>
        {' '}
        {subscribing
          ? '(Subscribing..)'
          : (
            <button onClick={this.toggleWebSocket} type="button">
              {subscribed ? 'Unsubscribe' : 'Subscribe'} Trades
            </button>
          )
        }
      </ActionRow>
    );
  }

  render() {
    return (
      <div>
        <HeaderRow>
          <p>Time</p>
          <p>Price</p>
          <p>Amount</p>
        </HeaderRow>
        <div>
          {this.renderRow()}
        </div>
        {this.renderActions()}
      </div>
    );
  }
}

Trades.propTypes = {
  trades: PropTypes.instanceOf(List), // eslint-disable-line
  initTrades: PropTypes.func.isRequired,
  updateTrades: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    trades: state.trades,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    initTrades: orders => dispatch(actionCreators.initTrades(orders)),
    updateTrades: order => dispatch(actionCreators.updateTrades(order)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Trades);
