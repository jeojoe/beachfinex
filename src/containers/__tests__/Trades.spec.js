import React from 'react';
import { shallow } from 'enzyme';
import { List } from 'immutable';

import { Trades } from '../Trades';

describe('<Trades />', () => {
  it('renders correctly', () => {
    const initTrades = jest.fn();
    const updateTrades = jest.fn();
    const trades = List([
      [3, 1547282908000, -1, 1],
      [2, 1547282907000, 1, 1],
    ]);
    const wrapped = shallow(
      <Trades
        initTrades={initTrades}
        bids={trades}
        updateTrades={updateTrades}
        ws={{
          subscribed: false,
          subscribing: true,
          setSubscribed: jest.fn(),
          subscribe: jest.fn(),
          unsubscribe: jest.fn(),
          toggle: jest.fn(),
          setMoreActions: jest.fn(),
        }}
      />,
    );
    expect(wrapped).toMatchSnapshot();
  });
});
