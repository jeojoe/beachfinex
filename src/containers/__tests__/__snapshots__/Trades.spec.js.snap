// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trades /> renders correctly 1`] = `
<div>
  <styled.div>
    <p>
      Time
    </p>
    <p>
      Price
    </p>
    <p>
      Amount
    </p>
  </styled.div>
  <div>
    Fetching..
  </div>
</div>
`;
